packages/pyright-internal/src/analyzer/typeEvaluator.ts

```diff
@@ -694,6 +696,28 @@ export function createTypeEvaluator(importLookup: ImportLookup, evaluatorOptions
         })?.type;
     }
 
+    // Determines the expected type of a specified node based on surrounding
+    // context. For example, if it's a subexpression of an argument expression,
+    // the associated parameter type might inform the expected type.
+    function getExpectedType(node: ExpressionNode): ExpectedTypeResult | undefined {
+        evaluateTypesForExpressionInContext(node);
+
+        let curNode: ParseNode | undefined = node;
+        while (curNode !== undefined) {
```

Comment:
so, I guess expectedType returns what code says on what type it expects and getType returns what type evaluator actually evaluated the type to?

basically, we have these functions that return type on expression node

getType: (node: ExpressionNode) => Type | undefined; 
getTypeOfExpression: (node: ExpressionNode, expectedType?: Type, flags?: EvaluatorFlags) => TypeResult;
getDeclaredTypeForExpression: (expression: ExpressionNode) => Type | undefined;

and wondering whether it can be merged to one of existing API? rather than consumer has to use right API at right location.

Reply:
getType and getExpectedType are the two main high-level APIs, and they return different things.

The others are lower-level APIs that I'd prefer not to expose outside of the type evaluator, but they're needed for some other tightly-bound modules like the checker.

Reply:
so what is difference between getType and getExpectedType? is my understanding correct?

Reply:
getType returns the evaluated type for that node. getExpectedType returns the type that is expected when using bidirectional inference. It is used in the process of evaluating the actual type.

For example:
```python
x: list[int] = ["a"]
```

In the case of the parse node for `"a"`, `getType` would return `Literal["a"]`, and `getExpectedType` would return `int`.